src/scene.rs

```diff
@@ -344,3 +347,51 @@ impl<'a> Iterator for Children<'a> {
     }
 }
 
+#[cfg(test)]
+mod tests {
+    use ::scene::Transform;
+
+    fn check(
+        t0: [f32; 3],
+        r0: [f32; 4],
+        s0: [f32; 3],
+        t1: [f32; 3],
+        r1: [f32; 4],
+        s1: [f32; 3],
+    ) {
+        assert_relative_eq!(t0[0], t1[0]);
+        assert_relative_eq!(t0[1], t1[1]);
+        assert_relative_eq!(t0[2], t1[2]);
+        assert_relative_eq!(r0[0], r1[0]);
+        assert_relative_eq!(r0[1], r1[1]);
+        assert_relative_eq!(r0[2], r1[2]);
+        assert_relative_eq!(r0[3], r1[3]);
+        assert_relative_eq!(s0[0], s1[0]);
+        assert_relative_eq!(s0[1], s1[1]);
+        assert_relative_eq!(s0[2], s1[2]);
+    }
+
+    #[test]
+    fn decompose_matrix_with_positive_determinant() {
+        let translation = [1.0, 2.0, 3.0];
+        let rotation = [0.0, 0.0, 0.0, 1.0];
+        let scale = [1.0, 1.0, 1.0];
+        let transform = Transform::Decomposed { translation, rotation, scale };
+        let matrix = transform.matrix();
+
+        let (t, r, s) = Transform::Matrix { matrix }.decomposed();
+        check(translation, rotation, scale, t, r, s);
+    }
+
+    #[test]
+    fn decompose_matrix_with_negative_determinant() {
+        let translation = [1.0, 2.0, 3.0];
+        let rotation = [0.0, 0.0, 0.0, 1.0];
+        let scale = [1.0, 1.0, -1.0];
```

Comment:
When the x or y scale is negative the decomposition is incorrect.

Reply:
Yeah, you can't check for equality of TRS components  because a TRS decomposition isn't unique; you have a choice of where to put the sign on the scale, you have a choice between q and -q for the rotation, etc. You could put `t`, `r`, and `s` into a new `Transform::Decomposed`, convert it to a matrix, and check for approximate equality with `matrix` instead.

Reply:
Round trip - I like that idea.

I wonder it's worth always picking the quaternion whose scalar value is positive, just so that decomposition is deterministic.

According to the specification, a negative z scale means the geometry must be rendered with the winding order reversed. It doesn't make too much sense to have a negative x or y scale as well, so perhaps we should ignore these cases and just make a note in the docs such as "matrices must not scale by a negative factor in the x or y axes".

Reply:
If you specify that the x and y scales are always positive and fix the quaternion you choose I think the decomposition is fully determined. For the quaternion you need to do slightly more than making the scalar positive because of the case when the scalar is zero. Something like  "the first component of the quaternion, counting from the end, that is non-zero is positive".

The new `test_decompose` looks good 👍 I think it'd be good to test a scale where the negative value is on the x or y component, since having it on the z component is kind of the "happy path" for `decomposed` and exercising the non-happy path would give me more confidence in the test.